### PR TITLE
Add ClusterRegistry support to operator

### PR DESCRIPTION
**Pull request by font**
This is a first stab at adding support to the federation-v2-operator for deploying the cluster registry. I would like to get some feedback before continuing further.

**NOTES**:
- Need to consider how best to manage the cluster registry CRD singleton using the operator when CRDs are cluster-scoped.
- This does not include instructions to deploy the cluster registry using the operator deployed with OLM, because I was not able to get OLM deployed successfully. So that work will be done in a follow-on PR.
- This doesn't include installing the cluster registry as a dependency of the federation-v2. I'm looking to add that support in a follow-on PR after getting an iteration of feedback for this work.
- We need to consider renaming this operator to the multi-cluster operator for a better UX.

### CODE REVIEW

File: cmd/federation-v2-operator/main.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ func main() {
 	if err != nil {
 		logrus.Fatalf("Failed to get watch namespace: %v", err)
 	}
-	resyncPeriod := 5
+	resyncPeriod := 0 // Disable periodic events i.e. only send events when updated.
```
Comment (marun) [inline cmd/federation-v2-operator/main.go]: What's the rational behind this change?

File: pkg/clusterregistry/handler.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	common.AddOwnerRefToObject(crNamespace, common.AsOwner(c, false))
+	return crNamespace
+}
+
+// clusterRegistryCRD returns the Cluster Registry CRD as an unstructured
+// object after downloading, converting, and adding the Cluster Registry
+// Operator CR object as its owner.
+func clusterRegistryCRD(c *v1alpha1.ClusterRegistry) (*unstructured.Unstructured, error) {
+	// TODO(font): what's the best way to get Cluster Registry CRD?
+	clusterRegistryURL := "https://raw.githubusercontent.com/kubernetes/cluster-registry/" + c.Spec.Version + "/cluster-registry-crd.yaml"
```
Comment (marun) [inline pkg/clusterregistry/handler.go]: Why should the crd be downloaded?  I was assuming that the operator would vendor in whatever it needed.

File: pkg/clusterregistry/handler.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	}
+	defer resp.Body.Close()
+
+	// Decode YAML
+	crd, err := readerToObj(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: Adding OwnerReference to CRD does not trigger garbage collector to
```
Comment (marun) [inline pkg/clusterregistry/handler.go]: So why bother with an ownerref?

File: pkg/clusterregistry/handler.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = createClusterRegistry(crObj)
+	if err != nil {
+		return err
+	}
+
+	// TODO(font): create necessary RBAC resources.
```
Comment (marun) [inline pkg/clusterregistry/handler.go]: Are you thinking a set of cluster roles that could be assigned to users (e.g. read_clusters, edit_clusters)?

File: pkg/clusterregistry/handler.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	// TODO(font): any status info to capture?
```
Comment (marun) [inline pkg/clusterregistry/handler.go]: Presumably an indication of whether the namespace and crd were installed successfully?  I haven't done that yet for deploying the control plane, but it is a logical step.

File: deploy/operator.yaml [hunk 6]
```diff
@@ -28,3 +28,11 @@ spec:
                   fieldPath: metadata.namespace
             - name: OPERATOR_NAME
               value: "federation-v2-operator"
+          # Mount host certs for verifying https connections
```
Comment (marun) [inline deploy/operator.yaml]: Why is this desirable or necessary?

**Comment by font** [comment]
@pmorie Yes, I agree. It's a workaround for a bug where owner references are not properly being honored in CRDs for deletion. So manual cleanup of the CRD (requiring the type from `kube-apiextensions-apiserver`) was necessary.

File: pkg/clusterregistry/handler.go (hunk 4 above)
Comment (font) [reply pkg/clusterregistry/handler.go]: Yep, cluster roles and/or roles. Need to think about that some more.

File: pkg/clusterregistry/handler.go (hunk 5 above)
Comment (font) [reply pkg/clusterregistry/handler.go]: Agreed.

File: pkg/clusterregistry/handler.go (hunk 2 above)
Comment (font) [reply pkg/clusterregistry/handler.go]: I didn't think it should be downloaded as it seemed fraught with danger. So I made a note to come back and address it. Yes, vendoring it in should be the way it's done. I had originally missed the definition of the cluster registry CRD in code so opted for this temporarily -- I've now discovered it.

File: pkg/clusterregistry/handler.go (hunk 3 above)
Comment (font) [reply pkg/clusterregistry/handler.go]: This seems like a bug to me and I've added a `BUG` comment above to investigate it further.

File: deploy/operator.yaml (hunk 6 above)
Comment (font) [reply deploy/operator.yaml]: It's not desirable, but necessary to download the CRD via https or use a different container image. But we should be able to remove the download and subsequently remove this.

File: cmd/federation-v2-operator/main.go (hunk 1 above)
Comment (font) [reply cmd/federation-v2-operator/main.go]: We keep getting the FederationV2 and ClusterRegistry CR object events every 5 seconds, even if there were no updates. That seems like unnecessary churn. If we disable it, we only get the updates whenever there was actually a change.

**Comment by font** [comment]
This is superseded by #3.

**Pull request closed without merging**